src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs

```diff
@@ -0,0 +1,273 @@
+
+                    _charBuffer = _charBuffer.Slice(overFlowChars);
+
+                    // readBuffer: [ 0, 0, ], overflowBuffer: [ 7, 13, 34, ]
+                    // Fill up the readBuffer to capacity, so the result looks like so:
+                    // readBuffer: [ 7, 13 ], overflowBuffer: [ 34 ]
+                    Debug.Assert(readBuffer.Count < overflowBytes);
+                    _overflowBuffer.Array.AsSpan(0, readBuffer.Count).CopyTo(readBuffer);
+
+                    Debug.Assert(_overflowBuffer.Array != null);
+
+                    _overflowBuffer = new ArraySegment<byte>(
+                        _overflowBuffer.Array,
+                        readBuffer.Count,
+                        overflowBytes - readBuffer.Count);
+
+                    Debug.Assert(_overflowBuffer.Count != 0);
+
+                    return readBuffer.Count;
+
+                default:
+                    Debug.Fail("We should never see this");
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private async Task ReadInputChars(CancellationToken cancellationToken)
+        {
+            Debug.Assert(_byteBuffer.Array != null);
+            // If we had left-over bytes from a previous read, move it to the start of the buffer and read content in to
+            // the segment that follows.
+            Buffer.BlockCopy(
+                _byteBuffer.Array,
+                _byteBuffer.Offset,
+                _byteBuffer.Array,
+                0,
+                _byteBuffer.Count);
+            int readBytes =
+#if NETSTANDARD2_0
+                await _stream.ReadAsync(_byteBuffer.Array, _byteBuffer.Count, _byteBuffer.Array.Length, cancellationToken).ConfigureAwait(false);
+#else
+                await _stream.ReadAsync(_byteBuffer.Array.AsMemory(_byteBuffer.Count), cancellationToken).ConfigureAwait(false);
+#endif
+            _byteBuffer = new ArraySegment<byte>(_byteBuffer.Array, 0, _byteBuffer.Count + readBytes);
+
+            Debug.Assert(_charBuffer.Count == 0, "We should only expect to read more input chars once all buffered content is read");
+
+#if NETSTANDARD2_0
+            _decoder.Convert(_byteBuffer.Array, _byteBuffer.Offset, _byteBuffer.Count, _charBuffer.Array, 0, _charBuffer.Array.Length,
+                flush: readBytes == 0, out int bytesUsed, out int charsUsed, out _);
+#else
+            _decoder.Convert(
+                _byteBuffer.AsSpan(),
+                _charBuffer.Array,
+                flush: readBytes == 0,
+                out int bytesUsed,
+                out int charsUsed,
+                out _);
+#endif
+
+            Debug.Assert(_charBuffer.Array != null);
+
+            _byteBuffer = _byteBuffer.Slice(bytesUsed);
+            _charBuffer = new ArraySegment<char>(_charBuffer.Array, 0, charsUsed);
+        }
+
+        private static void ThrowArgumentOutOfRangeException(byte[] buffer, int offset, int count)
```

Comment:
Nit: this is a strange name for this method; I'd expect ValidateArguments or something like that.  Generally methods that begin with "Throw" unconditionally throw, unless they have a conditional in their name, e.g. "ThrowXyzIfAbc".

Reply:
I moved this method's logic to the call site.